aten/src/ATen/native/cpu/SoftMaxKernel.cpp

```diff
@@ -207,6 +207,146 @@ struct vec_host_softmax_lastdim {
   }
 };
 
+#if defined(CPU_CAPABILITY_AVX2) && !defined(_MSC_VER)
+inline void _vec_softmax(
+    BFloat16* input_data_base,
+    BFloat16* output_data_base,
+    int64_t outer_size,
+    int64_t inner_size,
+    int64_t dim_size) {
+  using Vec = vec::Vectorized<float>;
+  using Vec_bf16 = vec::Vectorized<BFloat16>;
+  int64_t dim_stride = inner_size;
+  int64_t outer_stride = dim_size * dim_stride;
+  int64_t grain_size = std::min(internal::GRAIN_SIZE / dim_size, (int64_t)1);
+  int vectorized_step = Vec_bf16().size(); // Currently, we only support BFloat16 in this special implementation
+  TORCH_CHECK(
+    vectorized_step == 16,
+    "vectorized_step must be 16 with dtype bfloat16");
+  parallel_for(
+      0, outer_size * inner_size, grain_size, [&](int64_t begin, int64_t end) {
+        int64_t idx = begin;
+        std::unique_ptr<float[]> temp_scalar_input(new float[dim_size]());
```

Comment:
Why this over say [SmallVector](https://github.com/pytorch/pytorch/blob/master/c10/util/SmallVector.h)?

Reply:
@cpuhrsch Thanks for the suggestions. I am trying to use `SmallVector` as this `SmallVector<float, dim_size> temp_scalar_input();`. However, it can't pass the build phase since `SmallVector`  requires the second parameter `dim_size` to be a constant.

Reply:
And for the vectorized parts, `std::unique_ptr<float[]> temp_vec_input(new float[dim_size*vectorized_step*2]());` Since I want to [operate the data_ptr directly](https://github.com/pytorch/pytorch/pull/60371/files#diff-1d3fef70dbc64e3e716d056f635ed159f3b3bb161930bc126317c5ff11d88957R249), not sure how can I do it with `SmallVector`?

Reply:
Hm ok. How about using a Tensor? ```auto t = at::empty({dim_size}, torch::kFloat); float* temp_vec_input_data = t.data_ptr<float>()```. I just don't see much code like yours around the codebase, but let me ask some of my coworkers.